File: BraintreeModels/src/test/java/com/braintreepayments/api/ClientTokenUnitTest.java
package com.braintreepayments.api;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static com.braintreepayments.api.FixturesHelper.base64Encode;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNull;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
public class ClientTokenUnitTest {
    @Test
    public void fromString_deserializesClientToken() {
        ClientToken clientToken = (ClientToken) Authorization
                .fromString(base64Encode(Fixtures.CLIENT_TOKEN));

        assertEquals("client_api_configuration_url", clientToken.getConfigUrl());
        assertEquals("authorization_fingerprint", clientToken.getAuthorizationFingerprint());
    }

    @Test
    public void fromString_canDeserializeFromBase64String() {
        ClientToken clientToken = (ClientToken) Authorization.fromString(Fixtures.BASE64_CLIENT_TOKEN);

        assertEquals("encoded_capi_configuration_url", clientToken.getConfigUrl());
        assertEquals("encoded_auth_fingerprint", clientToken.getAuthorizationFingerprint());
    }

    @Test
    public void fromString_returnsInvalidTokenWhenGivenRandomJson() {
        Authorization result = ClientToken.fromString(Fixtures.RANDOM_JSON);

        assertTrue(result instanceof InvalidToken);
    }

    @Test
    public void getBearer_returnsAuthorizationFingerprint() {
        ClientToken clientToken = (ClientToken) Authorization
                .fromString(base64Encode(Fixtures.CLIENT_TOKEN));

        assertEquals(clientToken.getAuthorizationFingerprint(), clientToken.getBearer());
    }

    @Test
    public void getCustomerId_returnsNull_whenCustomerIdNotPresent() {
        ClientToken clientToken = (ClientToken) Authorization
                .fromString(base64Encode(Fixtures.CLIENT_TOKEN_WITH_AUTHORIZATION_FINGERPRINT_OPTIONS));

        assertNull(clientToken.getCustomerId());
    }

    @Test
    public void getCustomerId_returnsCustomerId() {
        ClientToken clientToken = (ClientToken) Authorization
                .fromString(base64Encode(Fixtures.CLIENT_TOKEN_WITH_CUSTOMER_ID_IN_AUTHORIZATION_FINGERPRINT));

        assertEquals("fake-customer-123", clientToken.getCustomerId());
    }
}

File: BraintreeCore/src/test/java/com/braintreepayments/api/ClientTokenUnitTest.kt
package com.braintreepayments.api

import com.braintreepayments.api.Authorization.Companion.fromString
import org.junit.Assert.*
import org.robolectric.RobolectricTestRunner
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(RobolectricTestRunner::class)
class ClientTokenUnitTest {
    @Test
    fun fromString_deserializesClientToken() {
        val clientToken =
            fromString(FixturesHelper.base64Encode(Fixtures.CLIENT_TOKEN)) as ClientToken
        assertEquals("client_api_configuration_url", clientToken.configUrl)
        assertEquals("authorization_fingerprint", clientToken.authorizationFingerprint)
    }

    @Test
    fun fromString_canDeserializeFromBase64String() {
        val clientToken = fromString(Fixtures.BASE64_CLIENT_TOKEN) as ClientToken
        assertEquals("encoded_capi_configuration_url", clientToken.configUrl)
        assertEquals("encoded_auth_fingerprint", clientToken.authorizationFingerprint)
    }

    @Test
    fun fromString_returnsInvalidTokenWhenGivenRandomJson() {
        val result = fromString(Fixtures.RANDOM_JSON)
        assertTrue(result is InvalidAuthorization)
    }

    @Test
    fun getBearer_returnsAuthorizationFingerprint() {
        val clientToken =
            fromString(FixturesHelper.base64Encode(Fixtures.CLIENT_TOKEN)) as ClientToken
        assertEquals(clientToken.authorizationFingerprint, clientToken.bearer)
    }

    @Test
    fun getCustomerId_returnsNull_whenCustomerIdNotPresent() {
        val clientToken =
            fromString(FixturesHelper.base64Encode(
                Fixtures.CLIENT_TOKEN_WITH_AUTHORIZATION_FINGERPRINT_OPTIONS)) as ClientToken
        assertNull(clientToken.customerId)
    }

    @Test
    fun getCustomerId_returnsCustomerId() {
        val clientToken =
            fromString(FixturesHelper.base64Encode(
                Fixtures.CLIENT_TOKEN_WITH_CUSTOMER_ID_IN_AUTHORIZATION_FINGERPRINT)) as ClientToken
        assertEquals("fake-customer-123", clientToken.customerId)
    }
}
